### ISSUE TITLE
Amiberry deletes simlinks when game starts and breakes kickstart references

### ISSUE THREAD

**Issue by tomino1983** [BlitterStudio/amiberry#1524]
**Describe the bug**
The issue happens when you try to start a game in lha format (WHDBoot) that requires kick34005.A500
It only starts if the simlink kick34005.A500 is located in one of the following folders:

- /tmp/amiberry/devs/Kickstarts
- /opt/retropie/configs/amiga/amiberry/whdboot/save-data/Kickstarts

When i launch amiberry it removes the link and on the screen i see the error:
DOS-Error #205 (object not found) on reading "devs:kickstarts/kick34005.a500

It doesnt happen with the real rom if i hard copy it in either one of the two folder listed up

**To Reproduce**

1. Create a simlink (ln -s ....) in either on of the identified folders

2. Launch a game

3. Check the existence of the link inside the foldere where it has been created

**Expected behavior**
I expect no need to copy the kickstart rom out of the standard /home/pi/RetroPie/BIOS/amiga folder in order to start the games that require it

**Screenshots**

**Desktop (please complete the following information):**

- OS: Raspbian Buster

- Version 10

**Additional context**
I think it started to behave like that after the latest update but i'm not so sure about it

Love Amiberry? Please consider supporting it:
👉  https://ko-fi.com/midwan

**Comment by giantclambake** [comment]
FWIW (testing with current master), I could not recreate this ~ all seems to be working as expected.... that said...

@midwan ~ correct me if I'm wrong, but `/tmp/amiberry` is only created at the time the whdbooter process is invoked ...ie; at whdload.lha load/start --- the `Rescan ROMs` button updates the content of (showing the path here for this setup) `~/.local/share/amiberry/whdboot/save-data/Kickstarts/*` ....when the whdbooter process launches, it creates the `/tmp/amiberry` directory and symlinks `/tmp/amiberry/devs/Kickstarts/*` to 'whdboot/save-data/Kickstarts/*'  at that time, not as a consequence of the `Rescan ROMs` button ...??... at least, that's how it works here in v7.0-RC1  ;)

**Comment by midwan** [comment]
@giantclambake 
Yes, that's how the process goes. But I want to be sure that `/tmp/amiberry` is clean first, before launching a game. I've seen some cases where the issue was related to that directory.

**Comment by giantclambake** [comment]
> I've seen some cases where the issue was related to that directory

Thanks...that'd be odd....something crashed before whdbooter did the clean-up?...does it test if `/tmp/amiberry` exists?...and if so, delete it and re-do from scratch?

**Comment by tomino1983** [comment]
I've just followed the steps and the tmp directory is created correctly but the required kickstart is not there nor symlinked.

I've notice that the Kickstarters folder is a symlink to retropi real folder:

```
pi@retropie:/tmp/amiberry $ ls -larth
total 0
drwxrwxrwt 11 root root 220 Dec  6 15:25 ..
drwxr-xr-x  5 pi   pi   100 Dec  6 15:25 .
drwxr-xr-x  2 pi   pi    60 Dec  6 15:25 s
drwxr-xr-x  2 pi   pi    60 Dec  6 15:25 devs
drwxr-xr-x  2 pi   pi    60 Dec  6 15:25 c
pi@retropie:/tmp/amiberry $ cd devs/
pi@retropie:/tmp/amiberry/devs $ ls -larth
total 0
drwxr-xr-x 5 pi pi 100 Dec  6 15:25 ..
lrwxrwxrwx 1 pi pi  61 Dec  6 15:25 Kickstarts -> /opt/retropie/emulators/amiberry/whdboot/save-data/Kickstarts
drwxr-xr-x 2 pi pi  60 Dec  6 15:25 .
pi@retropie:/tmp/amiberry/devs $
```

the games start only if i put kick34005.A500 inside that folder, but as i write it only contains:
```
pi@retropie:/tmp/amiberry/devs $ ls Kickstarts
foo.txt             kick33192.A500.RTB  kick40063.A600.RTB   kick40068.A4000.RTB
kick33180.A500.RTB  kick34005.A500.RTB  kick40068.A1200.RTB
pi@retropie:/tmp/amiberry/devs $
```

Thank you all

Tommaso

**Comment by midwan** [comment]
@tomino1983 
Thanks for the feedback.
Can you give me some titles (at least one) that requires the kickstart file, so I can try to recreate the issue here?

**Comment by tomino1983** [comment]
Mainly two tiles, Ugh! and TURBO:

Ugh_v1.2_0396.lha

Turbo_v1.1_3087.lha

One other feedback; I've double checked but after i select "Rescan Paths" nothing happens at filesystem level, i mean i don't see the symlinks inside the /opt/retropie/emulators/amiberry/whdboot/save-data/Kickstarts but i'm not really sure that i'm supposed to see them there.

Thank you

**Comment by giantclambake** [comment]
Both titles work as expected here ~ it seems to me as though the symlinks are not created at all  (in `/opt/retropie/emulators/amiberry/whdboot/save-data/Kickstarts`) when the `Rescan Paths` button is pressed....ie; amiberry isn't deleting anything in `/tmp/amiberry` ~ the symlinks don't exist when whdbooter starts.

**Comment by midwan** [comment]
@giantclambake 
Where did you get the lowercase filename from?
The symlinks are created with capital "A", and should point to the ROM files (if found).

For example, this is what I get in my local Debian installation, after running `Ugh!`:

![image](https://github.com/user-attachments/assets/01dfd408-1c8b-4a10-bea5-b7634e0a35bc)

**Comment by tomino1983** [comment]
Thats what i'm talking about. I totally miss the symlinks and if i create them manually as soon as i launch amiberry it deletes them. The question now is why all the other games a working properly without symlinks of the other kickstarts as well.

**Comment by giantclambake** [comment]
@midwan  ~ it's subtle, I'm referring to the filenames of the rom themselves in rom_path ... umm....

`cd ~/Amiberry/roms`
`cp amiga-os-130.rom kick34005.a500`   //identical images, different name

Launch amiberry -> GUI ->Paths panel --> Rescan Paths button --> Ok ---> Quit amiberry

resultant symlink:

~/.local/share/amiberry/whdboot/save-data/Kickstarts/kick34005.A500 --> /home/gcb/Amiberry/roms/kick34005.a500

`mv kick34005.a500 kick34005.A500`

Launch amiberry -> GUI ->Paths panel --> Rescan Paths button --> Ok ---> Quit amiberry

resultant symlink:

~/.local/share/amiberry/whdboot/save-data/Kickstarts/kick34005.A500 --> /home/gcb/Amiberry/roms/amiga-os-130.rom

`mv amiga-os-130.rom ~/`

Launch amiberry -> GUI ->Paths panel --> Rescan Paths button --> Ok ---> Quit amiberry

The symlink ~/.local/share/amiberry/whdboot/save-data/Kickstarts/kick34005.A500 **is not created**

`mv kick34005.A500 kick34005.a500`

Launch amiberry -> GUI ->Paths panel --> Rescan Paths button --> Ok ---> Quit amiberry

resultant symlink:

~/.local/share/amiberry/whdboot/save-data/Kickstarts/kick34005.A500 --> /home/gcb/Amiberry/roms/kick34005.a500

*I noticed @tomino1983 kept mentioning the rom filename being copied was kick34005.A500 ...with a capital 'A' ...

Better?  ;)

**Comment by midwan** [comment]
@giantclambake 
That is not related to this case's problem though, it has to do with Amiberry's pattern recognition for ROM files, which get added to the cache. But in this case, the symlinks are meant to be created in the `/tmp/amiberry/devs/Kickstarts` directory, for use with WHDLoad inside the emulated environment. 

Nevertheless, I've just pushed a commit that improves that, and now capitalized extensions will be picked up by the scanner also.

Regarding this case, I suspect that something is going wrong when trying to generate the symlinks, during the WHDLoad booter process. The normal procedure is:
- If the symlinks are found, they are removed (what @tomino1983 experiences)
- Then there's a check if the symlinks are missing (which they should, considering they were just removed), and it will try to generate them based on the rom file (the filename doesn't matter here, it will detect the ROM file and create a symlink to it).
- If creating the symlink fails, it will try to make a copy of the file, instead.

So, if it's failing to generate the symlinks, I suspect that either the symlink generation/copy file operation are failing, or that the required romfile it should symlink to, is not recognized.

@tomino1983 please enable logging and see if we can get some more info there, as there should be relevant log entries in case the symlink generation / copy fails. Make sure you point the logfile to a directory Amiberry can write to!

**Comment by giantclambake** [comment]
@midwan  --- I surely suspect that commit fixes things here, but to elucidate further...

> That is not related to this case's problem though, it has to do with Amiberry's pattern recognition for ROM files, which get added to the cache.

I beg to differ, as exampled ~ Amiberry's pattern recognition does **not detect** the rom filename `kick34005.A500`

Test case: (before bef8f60)

```
gcb@gallah:~$ ls ~/Amiberry/roms
Kick34005.A500
```

Launch amiberry -> GUI ->Paths panel --> Rescan Paths button .....

result: Rescan Paths button remains shaded (active) for a couple of 10s of seconds, then amiberry GUI quietly closes. (logfile attached)

![ex](https://imgur.com/ty27xTt.png)

[amiberry.log.gz](https://github.com/user-attachments/files/18050226/amiberry.log.gz)


**Big Fat Note**: As you can see in the above logfile, the reason it sat there for a couple of 10s of seconds, was it decided to recursively search my $HOMEDIR looking for...ummm..."some kind of match", I guess.... and I see it scanned the directory `/home/gcb/Downloads/rombak` ...which does indeed contain backup copies of my Amiga kickstart roms...it crashed shortly thereafter ;) At some level, this is 'bad behaviour' ---- amiberry should **NOT** search for rom files outside of $rom_path, end of story.

In this situation;

```
gcb@gallah:~$ amiberry -G
terminate called after throwing an instance of 'gcn::Exception'
Aborted (core dumped)
```
Note: result is a corollary of this test case --> AROS roms are not present, and the (ostensibly valid) Kick34005.A500 kickstart rom file has not been recognized in scanroms, nor entered into cache...

....and obviously the cited whdload.lha titles will not run....

...following on, @tomino1983 indicated that the file `Kick34005.A500` was being copied (not symlinked) into the `/tmp/amiberry/devs/Kickstarts/`  location.

As far as amiberry is concerned, **the cache structure has been 'set'**, and between starts it examines `/tmp/amiberry/devs/Kickstarts/`, finds the filename `Kick34005.A500` and dutifully deletes it (not part of cached structure, doesn't care if it's a symlink or not) --- it cannot create the appropriate symlink here (because...) the rom filename `Kick34005.A500` has not been detected nor cached.

--------------------
...leaving things 'as is', and retesting with amiberry after commit bef8f60 ....
--------------------

Launch amiberry -> GUI ->Paths panel --> Rescan Paths button --> Ok ---> Quit amiberry  //works as expected

gcb@gallah:~$ amiberry -G   //works as expected, booting to ks1.3 insert disk screen

![ex2](https://imgur.com/FIR42zF.png)

Direct-loading  of  Ugh_v1.2_0396.lha and/or Turbo_v1.1_3087.lha will still fail at this point, due to the fact I only have a single rom image in ~/Amiberry/roms  for this test...

`WHDBooter - Host: A600 ROM not found`

Note: once all the required rom files for the whdbooter are present, and you Rescan Paths, these titles launch & run as expected.

The point of this test, was to try recreate what @tomino1983 described above...ie; "_One other feedback; I've double checked but after i select "Rescan Paths" nothing happens at filesystem level, i mean i don't see the symlinks inside the /opt/retropie/emulators/amiberry/whdboot/save-data/Kickstarts but i'm not really sure that i'm supposed to see them there._"

I could recreate this ~ I've explained how (even on a non-retropie setup). To me, it seems _very much_ related to this issue, but as I say the problem is/was subtle ;) I did not check other Amiga romfiles/names, just kick34005.A500 (other titles working fine reportedly).

> Nevertheless, I've just pushed a commit that improves that, and now capitalized extensions will be picked up by the scanner also.

Kudos, in my testing here (no retropie environment), suggest that fixes this issue ~ caveat: needs to be tested in retropie
This would be..what?...the 2nd or 3rd time 'case insensitivity' has bitten us after the WinUAE merges? I'll choose to be expecting of other machinations of this type, all things considered 8)

@tomino1983 ~ if possible, I would _advise_ rebuilding amiberry from a git pull of the current master tree, and retesting 8)
Or, alternately, you can test the hypothesis by doing something like...

```
cd /home/pi/RetroPie/BIOS/amiga
cp kick34005.A500 kick34005.a500
```

Launch amiberry -> GUI ->Paths panel --> Rescan Paths button --> Ok  ...check results or retest whdload.lha titles mentioned.

**Comment by tomino1983** [comment]
Man. thank you, you are totally right.

I've just copied A500 to a500, rescanned paths, and this is the result:

`pi@retropie:/tmp/amiberry/devs/Kickstarts $ ls -larth
total 48K
drwxr-xr-x 6 pi pi 4.0K Jan 21  2023 ..
-rwxr-xr-x 1 pi pi 4.9K Dec  6 13:38 kick40068.A4000.RTB
-rwxr-xr-x 1 pi pi 4.9K Dec  6 13:38 kick40068.A1200.RTB
-rwxr-xr-x 1 pi pi 4.9K Dec  6 13:38 kick40063.A600.RTB
-rwxr-xr-x 1 pi pi 4.0K Dec  6 13:38 kick34005.A500.RTB
-rwxr-xr-x 1 pi pi 3.8K Dec  6 13:38 kick33192.A500.RTB
-rwxr-xr-x 1 pi pi 3.8K Dec  6 13:38 kick33180.A500.RTB
-rw-r--r-- 1 pi pi    1 Dec  6 13:38 foo.txt
lrwxrwxrwx 1 pi pi   58 Dec  8 12:40 kick34005.A500 -> /opt/retropie/emulators/amiberry/kickstarts/kick34005.a500
drwxr-xr-x 2 pi pi 4.0K Dec  8 12:40 .
pi@retropie:/tmp/amiberry/devs/Kickstarts $`


But the problem now is that everytime amiberry finds a lowercase "a" in the kickstarts name it throws the error : 
`Unknow` command FAILAT
1`

So the sysmlink is created an not delated when i start UGH (for example). But it seems that the presence of the same kickstart with the little "a" breaks it even more.

During the day i'll compile the last version and test it.

One more point. If i do the same test for the others kickstarters i have the following behavior:
`pi@retropie:/tmp/amiberry/devs/Kickstarts $ ls -lart /home/pi/RetroPie/BIOS/amiga/
total 6160
-rw-r--r-- 1 pi pi 524288 Aug  6  2008 kick40068.A4000
-rw-r--r-- 1 pi pi 524288 Aug  6  2008 kick40068.A1200
-rw-r--r-- 1 pi pi 524288 Oct 25  2020 kick40063.A600
-rw-r--r-- 1 pi pi 262144 Sep  9  2022 kick33180.A500
-rw-r--r-- 1 pi pi 524288 Sep 12  2022 kick37350.A600
drwxr-xr-x 2 pi pi   4096 Oct  4 03:26 mt32-roms
-rw-r--r-- 1 pi pi    135 Oct  4 03:26 dir.txt
-rw-r--r-- 1 pi pi 524288 Oct  4 03:26 aros-rom.bin
-rw-r--r-- 1 pi pi 524288 Oct  4 03:26 aros-ext.bin
-rw-r--r-- 1 pi pi 262144 Dec  6 01:31 kick34005.A500
drwxr-xr-x 9 pi pi   4096 Dec  6 13:37 ..
-rw-r--r-- 1 pi pi 262144 Dec  8 12:39 kick34005.a500
-rw-r--r-- 1 pi pi 262144 Dec  8 12:53 kick33180.a500
-rw-r--r-- 1 pi pi 524288 Dec  8 12:53 kick37350.a600
-rw-r--r-- 1 pi pi 524288 Dec  8 12:53 kick40063.a600
-rw-r--r-- 1 pi pi 524288 Dec  8 12:53 kick40068.a1200
drwxr-xr-x 3 pi pi   4096 Dec  8 12:54 .
-rw-r--r-- 1 pi pi 524288 Dec  8 12:54 kick40068.a4000
pi@retropie:/tmp/amiberry/devs/Kickstarts $ ls -lart /tmp/amiberry/devs/Kickstarts/
total 48
drwxr-xr-x 6 pi pi 4096 Jan 21  2023 ..
-rwxr-xr-x 1 pi pi 4976 Dec  6 13:38 kick40068.A4000.RTB
-rwxr-xr-x 1 pi pi 5008 Dec  6 13:38 kick40068.A1200.RTB
-rwxr-xr-x 1 pi pi 5000 Dec  6 13:38 kick40063.A600.RTB
-rwxr-xr-x 1 pi pi 4020 Dec  6 13:38 kick34005.A500.RTB
-rwxr-xr-x 1 pi pi 3882 Dec  6 13:38 kick33192.A500.RTB
-rwxr-xr-x 1 pi pi 3882 Dec  6 13:38 kick33180.A500.RTB
-rw-r--r-- 1 pi pi    1 Dec  6 13:38 foo.txt
lrwxrwxrwx 1 pi pi   59 Dec  8 12:54 kick40068.A1200 -> /opt/retropie/emulators/amiberry/kickstarts/kick40068.a1200
lrwxrwxrwx 1 pi pi   58 Dec  8 12:54 kick34005.A500 -> /opt/retropie/emulators/amiberry/kickstarts/kick34005.a500
lrwxrwxrwx 1 pi pi   58 Dec  8 12:54 kick33180.A500 -> /opt/retropie/emulators/amiberry/kickstarts/kick33180.a500
drwxr-xr-x 2 pi pi 4096 Dec  8 12:54 .
pi@retropie:/tmp/amiberry/devs/Kickstarts $`

It means that amiberry creates the symlink for all the other kickstartes but A600 and A4000, that is maybe the reason why other games work beacuse they use those skipped kickstartes versions.

Thank you

Thank you very much

**Comment by giantclambake** [comment]
@tomino1983  -- thanks for checking ~ see how a build from current `master` tree goes for you.

....I should point out, those filenames for the rom images are in fact part of the catalyst here ...for example;

![ex](https://imgur.com/tB2dNfk.png)

...at a guess, midwan's setup will be similar (iirc these are the default names from AmigaForever distrib), and likewise if you read https://github.com/BlitterStudio/amiberry/wiki/Kickstart-ROMs-(BIOS), the example name there for this specific rom image is "Kickstart v1.3 rev 34.5 (1987)(Commodore)(A500-A1000-A2000-CDTV).rom" .... see what I mean? ;)

...specifically it's the .A500/.a500 filename extension that's problematic here -- those extensions are used by the whdbooter process when symlinking the required Amiga kickstart images --- generally speaking, the kickstart images themselves, are filename.rom or filename.bin whatever...but _NOT_  filename.A500 or filename.a500 ....

...not saying your fault, just pointing out how the choice of Amiga kickstart filename ended up unraveling things here ;)

**Comment by midwan** [comment]
I don't currently have a RetroPie installation here to test with directly but also keep in mind that the version they use there is _not_ the same as the latest found on this repository. RetroPie uses v5.x which is now separated out into the Amiberry-Lite repository, but it hasn't yet been updated with all the latest changes going into v7, which is found in this repo. 

That will happen after v7.0 final is out, which is hopefully not too far now (we currently have RC1 out for testing).

In other words, my changes above are not found in Amiberry-Lite (v5) yet.

**Comment by giantclambake** [comment]
Doh, my bad... I should've realized we're referencing amiberry-lite here, and not amiberry.....(and this becoming a small pita btw....that is, amiberry-lite issues ending up here)....but not to worry...

Incidentally, the 'quick fix' here really, was to rename all the Amiga kickstart files, to have the more usual .bin/.rom extension ..ie; mv kick34005.A500 kick34005.rom  ;)

**Comment by giantclambake** [comment]
@midwan 
Ahh, that I wasn't aware of ~ right you are then ;) Ergo.. it does seem to stem from the case insensitivity from that other very well known OS....

...nah, I think the amiberry wiki seems fine ...I'll read through it again just to be sure ~ if there were any 'blame' to assign here, I'd say it's the table that appears on the page https://retropie.org.uk/docs/Amiga/ , but that's inconsequential now, as we handle those filename extensions 8)

**Comment by midwan** [comment]
@tomino1983 
Did you get a chance to test this yet?

**Comment by midwan** [comment]
OK, I will assume this is fixed for now, until told otherwise.
Keep in mind these changes are only in this repo's version for now, but will eventually (after v7 final is out) be back-ported to Amiberry-lite as well.

**Comment by tomino1983** [comment]
> 
> 
> @tomino1983 ~ Tell us, what exactly gave you the impression to rename the Amiga kickstart images to be **kick34005.A500** , **kick40063.A600**, **kick40068.A1200**....etc etc...??

The retropi guide you linked.

@midwan i'm out of home so i have gotten the chance to test it, but i'll let you know as soon as i do it.

Thank you all for the support.
